# Import datasets

In [1]:
#Import pandas, and read the train file
import pandas as pd

train = pd.read_csv('train.csv', index_col="Id")

In [2]:
#Check how many cols and rows the data has
print(train.shape)

(1460, 80)


In [3]:
#Check the first five rows of the data
train.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
4          Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
5          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
1        0      2    2008        WD         Normal     208500  
2        0      5    2007        WD         Normal     181500  
3        0      9    2008        WD         Normal     223500  
4        0      2    2006        WD        Abnorml     140000  
5        0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [4]:
#Check a summary of the data
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1460 non-null   int64  
 1   MSZoning       1460 non-null   object 
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   Street         1460 non-null   object 
 5   Alley          91 non-null     object 
 6   LotShape       1460 non-null   object 
 7   LandContour    1460 non-null   object 
 8   Utilities      1460 non-null   object 
 9   LotConfig      1460 non-null   object 
 10  LandSlope      1460 non-null   object 
 11  Neighborhood   1460 non-null   object 
 12  Condition1     1460 non-null   object 
 13  Condition2     1460 non-null   object 
 14  BldgType       1460 non-null   object 
 15  HouseStyle     1460 non-null   object 
 16  OverallQual    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  YearBuil

# Explore The Data

## Number Of Missing Values By Column

In [5]:
#Count (using `.sum()`) the number of missing values (`.isnull()`) in each column of 
#the train data as well as the percentages (using `.mean()` instead of `.sum()`).
#Order them (increasing or decreasing) using sort_values
#Call `pd.concat` to present these in a single table (DataFrame) with the helpful column names 'count' and '%'
missing = pd.concat([train.isnull().sum(), 100 * train.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='%', ascending=False)

count          %
PoolQC        1453  99.520548
MiscFeature   1406  96.301370
Alley         1369  93.767123
Fence         1179  80.753425
FireplaceQu    690  47.260274
...            ...        ...
Heating          0   0.000000
HeatingQC        0   0.000000
MSZoning         0   0.000000
1stFlrSF         0   0.000000
SalePrice        0   0.000000

[80 rows x 2 columns]

In [6]:
#List missing values only
missing.loc[missing['count']>0].sort_values(by='%', ascending=False)

count          %
PoolQC         1453  99.520548
MiscFeature    1406  96.301370
Alley          1369  93.767123
Fence          1179  80.753425
FireplaceQu     690  47.260274
LotFrontage     259  17.739726
GarageType       81   5.547945
GarageYrBlt      81   5.547945
GarageFinish     81   5.547945
GarageQual       81   5.547945
GarageCond       81   5.547945
BsmtExposure     38   2.602740
BsmtFinType2     38   2.602740
BsmtFinType1     37   2.534247
BsmtCond         37   2.534247
BsmtQual         37   2.534247
MasVnrArea        8   0.547945
MasVnrType        8   0.547945
Electrical        1   0.068493

### Columns having more than 80% of missing values

I will drop the columns from 'PoolQC' to 'Fence' as they have more than 80% of missing values and these will not work efficiently on deciding prices. 

In [7]:
train = train.drop(['PoolQC', 'MiscFeature', 'Alley', 'Fence'], axis=1)

In [8]:
print(train.shape)

(1460, 76)


Columns are reduced from (1460, 80) to (1460, 76)

In [9]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 76 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1460 non-null   int64  
 1   MSZoning       1460 non-null   object 
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   Street         1460 non-null   object 
 5   LotShape       1460 non-null   object 
 6   LandContour    1460 non-null   object 
 7   Utilities      1460 non-null   object 
 8   LotConfig      1460 non-null   object 
 9   LandSlope      1460 non-null   object 
 10  Neighborhood   1460 non-null   object 
 11  Condition1     1460 non-null   object 
 12  Condition2     1460 non-null   object 
 13  BldgType       1460 non-null   object 
 14  HouseStyle     1460 non-null   object 
 15  OverallQual    1460 non-null   int64  
 16  OverallCond    1460 non-null   int64  
 17  YearBuilt      1460 non-null   int64  
 18  YearRemo

### FireplaceQu

In case of 'FieplaceQu', almost half are missing values, and the null actually means 'No Fireplace', which is already shown in 'Fireplaces' column, so I will drop it as well.

In [10]:
train['Fireplaces'].value_counts()

0    690
1    650
2    115
3      5
Name: Fireplaces, dtype: int64

In [11]:
train['FireplaceQu'].value_counts()

Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64

In [12]:
fq = train.loc[train['FireplaceQu'].isnull()]
fq[['FireplaceQu', 'Fireplaces']]

FireplaceQu  Fireplaces
Id                          
1            NaN           0
6            NaN           0
11           NaN           0
13           NaN           0
16           NaN           0
...          ...         ...
1453         NaN           0
1454         NaN           0
1455         NaN           0
1459         NaN           0
1460         NaN           0

[690 rows x 2 columns]

In case of 'FieplaceQu', almost half are missing values so I will drop it as well.
And these null values actually mean 'No Fireplace', which is same as 0 in 'Fireplaces' column.

In [13]:
train = train.drop(['FireplaceQu'], axis=1)

In [14]:
print(train.shape)

(1460, 75)


Columns are reduced from (1460, 76) to (1460, 75)

### LotFrontage

In case of 'LotFrontage', I may can fill out with mean values of the column.

In [15]:
train['LotFrontage'].max(), train['LotFrontage'].min(), train['LotFrontage'].mean()

(313.0, 21.0, 70.04995836802665)

However, 17.7% of the column are null and it looks dropping would be better.

In [16]:
train = train.drop(['LotFrontage'], axis=1)

In [17]:
print(train.shape)

(1460, 74)


Columns are reduced from (1460, 75) to (1460, 74)

### Garage related columns

In [18]:
#Check if there is any abnormal data in 'GarageType' column
train['GarageType'].value_counts(), train['GarageType'].value_counts().sum()

(Attchd     870
 Detchd     387
 BuiltIn     88
 Basment     19
 CarPort      9
 2Types       6
 Name: GarageType, dtype: int64,
 1379)

GarageType: Garage location
		
       2Types	More than one type of garage
       Attchd	Attached to home
       Basment	Basement Garage
       BuiltIn	Built-In (Garage part of house - typically has room above garage)
       CarPort	Car Port
       Detchd	Detached from home
       NA	No Garage

In [19]:
#Check if there is any abnormal data in 'GarageYrBlt' column
train['GarageYrBlt'].value_counts(), train['GarageYrBlt'].value_counts().sum()

(2005.0    65
 2006.0    59
 2004.0    53
 2003.0    50
 2007.0    49
           ..
 1908.0     1
 1927.0     1
 1933.0     1
 1900.0     1
 1906.0     1
 Name: GarageYrBlt, Length: 97, dtype: int64,
 1379)

In [20]:
#Check if there is any abnormal data in 'GarageFinish' column
train['GarageFinish'].value_counts(), train['GarageFinish'].value_counts().sum()

(Unf    605
 RFn    422
 Fin    352
 Name: GarageFinish, dtype: int64,
 1379)

GarageFinish: Interior finish of the garage

       Fin	Finished
       RFn	Rough Finished	
       Unf	Unfinished
       NA	No Garage

In [21]:
#Check if there is any abnormal data in 'GarageQual' column
train['GarageQual'].value_counts(), train['GarageQual'].value_counts().sum()

(TA    1311
 Fa      48
 Gd      14
 Po       3
 Ex       3
 Name: GarageQual, dtype: int64,
 1379)

In [22]:
#Check if there is any abnormal data in 'GarageCond' column
train['GarageCond'].value_counts(), train['GarageCond'].value_counts().sum()

(TA    1326
 Fa      35
 Gd       9
 Po       7
 Ex       2
 Name: GarageCond, dtype: int64,
 1379)

GarageQual: Garage quality

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
		
GarageCond: Garage condition

       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage

The null values actually would mean "No Garage", not really missing data.

In addition, we can assume that the columns 'Garage*' would have null for the same rows, let's double-check in another way if it is true.

In [23]:
garage_cols = ['GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond']
n = train.loc[train[garage_cols].isnull().all(1)]
n[garage_cols].shape

(81, 5)

In [24]:
gt = train.loc[train['GarageType'].isnull()]
gt['GarageType'].shape

(81,)

In [25]:
gf = train.loc[train['GarageFinish'].isnull()]
gf['GarageFinish'].shape

(81,)

In [26]:
gy = train.loc[train['GarageYrBlt'].isnull()]
gy['GarageYrBlt'].shape

(81,)

In [27]:
gq = train.loc[train['GarageQual'].isnull()]
gq['GarageQual'].shape

(81,)

In [28]:
gc = train.loc[train['GarageCond'].isnull()]
gc['GarageCond'].shape

(81,)

We could see that the null values over the 5 columns appear in one row.
I will replace them to "NoGrg" instead of dropping them, except 'GarageYrBlt' to be 0.
(I don't think it's a good idea to set null in 'GarageYrBlt' to 0, but have no idea for now what would be the best..)

In [29]:
train = train.fillna({'GarageType':'NoGrg', 'GarageYrBlt':0, 'GarageFinish':'NoGrg', 'GarageQual':'NoGrg', 'GarageCond':'NoGrg'})

In [30]:
garage_cols = ['GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond']
n = train.loc[train[garage_cols].isnull().all(1)]
n[garage_cols].shape

(0, 5)

Null is all gone in all 5 columns.

### Basement related columns

In [31]:
#Check if there is any abnormal data in 'BsmtExposure' column
train['BsmtExposure'].value_counts(), train['BsmtExposure'].value_counts().sum()

(No    953
 Av    221
 Gd    134
 Mn    114
 Name: BsmtExposure, dtype: int64,
 1422)

In [32]:
#Check if there is any abnormal data in 'BsmtFinType2' column
train['BsmtFinType2'].value_counts(), train['BsmtFinType2'].value_counts().sum()

(Unf    1256
 Rec      54
 LwQ      46
 BLQ      33
 ALQ      19
 GLQ      14
 Name: BsmtFinType2, dtype: int64,
 1422)

In [33]:
#Check if there is any abnormal data in 'BsmtFinType1' column
train['BsmtFinType1'].value_counts(), train['BsmtFinType1'].value_counts().sum()

(Unf    430
 GLQ    418
 ALQ    220
 BLQ    148
 Rec    133
 LwQ     74
 Name: BsmtFinType1, dtype: int64,
 1423)

In [34]:
#Check if there is any abnormal data in 'BsmtCond' column
train['BsmtCond'].value_counts(), train['BsmtCond'].value_counts().sum()

(TA    1311
 Gd      65
 Fa      45
 Po       2
 Name: BsmtCond, dtype: int64,
 1423)

In [35]:
#Check if there is any abnormal data in 'BsmtQual' column
train['BsmtQual'].value_counts(), train['BsmtQual'].value_counts().sum()

(TA    649
 Gd    618
 Ex    121
 Fa     35
 Name: BsmtQual, dtype: int64,
 1423)

In [36]:
#Create a list of Basement related columns that have null
bsmt_cols = ['BsmtExposure', 'BsmtFinType2', 'BsmtFinType1', 'BsmtCond', 'BsmtQual']

#Create a list of all basement related columns
bsmt_all_cols = ['BsmtExposure', 'BsmtFinType2', 'BsmtFinType1', 'BsmtCond', 'BsmtQual','BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF']

In [37]:
n = train.loc[train[bsmt_cols].isnull().all(1)]
n[bsmt_all_cols].shape

(37, 9)

In [38]:
n[bsmt_all_cols]

BsmtExposure BsmtFinType2 BsmtFinType1 BsmtCond BsmtQual  BsmtFinSF1  \
Id                                                                          
18            NaN          NaN          NaN      NaN      NaN           0   
40            NaN          NaN          NaN      NaN      NaN           0   
91            NaN          NaN          NaN      NaN      NaN           0   
103           NaN          NaN          NaN      NaN      NaN           0   
157           NaN          NaN          NaN      NaN      NaN           0   
183           NaN          NaN          NaN      NaN      NaN           0   
260           NaN          NaN          NaN      NaN      NaN           0   
343           NaN          NaN          NaN      NaN      NaN           0   
363           NaN          NaN          NaN      NaN      NaN           0   
372           NaN          NaN          NaN      NaN      NaN           0   
393           NaN          NaN          NaN      NaN      NaN           0   
521           NaN          NaN          NaN      NaN      NaN           0   
533           NaN          NaN          NaN      NaN      NaN           0   
534           NaN          NaN          NaN      NaN      NaN           0   
554           NaN          NaN          NaN      NaN      NaN           0   
647           NaN          NaN          NaN      NaN      NaN           0   
706           NaN          NaN          NaN      NaN      NaN           0   
737           NaN          NaN          NaN      NaN      NaN           0   
750           NaN          NaN          NaN      NaN      NaN           0   
779           NaN          NaN          NaN      NaN      NaN           0   
869           NaN          NaN          NaN      NaN      NaN           0   
895           NaN          NaN          NaN      NaN      NaN           0   
898           NaN          NaN          NaN      NaN      NaN           0   
985           NaN          NaN          NaN      NaN      NaN           0   
1001          NaN          NaN          NaN      NaN      NaN           0   
1012          NaN          NaN          NaN      NaN      NaN           0   
1036          NaN          NaN          NaN      NaN      NaN           0   
1046          NaN          NaN          NaN      NaN      NaN           0   
1049          NaN          NaN          NaN      NaN      NaN           0   
1050          NaN          NaN          NaN      NaN      NaN           0   
1091          NaN          NaN          NaN      NaN      NaN           0   
1180          NaN          NaN          NaN      NaN      NaN           0   
1217          NaN          NaN          NaN      NaN      NaN           0   
1219          NaN          NaN          NaN      NaN      NaN           0   
1233          NaN          NaN          NaN      NaN      NaN           0   
1322          NaN          NaN          NaN      NaN      NaN           0   
1413          NaN          NaN          NaN      NaN      NaN           0   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
Id                                        
18             0          0            0  
40             0          0            0  
91             0          0            0  
103            0          0            0  
157            0          0            0  
183            0          0            0  
260            0          0            0  
343            0          0            0  
363            0          0            0  
372            0          0            0  
393            0          0            0  
521            0          0            0  
533            0          0            0  
534            0          0            0  
554            0          0            0  
647            0          0            0  
706            0          0            0  
737            0          0            0  
750            0          0            0  
779            0          0            0  
869            0          0            0  
895         

In [39]:
#Check how many rows exist when both 'BsmtExposure' and 'BsmtFinType2' have null
n2 = train.loc[train[['BsmtExposure', 'BsmtFinType2']].isnull().all(1)]
n2[bsmt_cols].shape

(37, 5)

I expected I would get 38 rows, but got 37. This means that not the all null rows doesn't go along with this combination.

In [40]:
n3 = train.loc[train[['BsmtExposure']].isnull().all(1)]
n3[bsmt_all_cols]

BsmtExposure BsmtFinType2 BsmtFinType1 BsmtCond BsmtQual  BsmtFinSF1  \
Id                                                                          
18            NaN          NaN          NaN      NaN      NaN           0   
40            NaN          NaN          NaN      NaN      NaN           0   
91            NaN          NaN          NaN      NaN      NaN           0   
103           NaN          NaN          NaN      NaN      NaN           0   
157           NaN          NaN          NaN      NaN      NaN           0   
183           NaN          NaN          NaN      NaN      NaN           0   
260           NaN          NaN          NaN      NaN      NaN           0   
343           NaN          NaN          NaN      NaN      NaN           0   
363           NaN          NaN          NaN      NaN      NaN           0   
372           NaN          NaN          NaN      NaN      NaN           0   
393           NaN          NaN          NaN      NaN      NaN           0   
521           NaN          NaN          NaN      NaN      NaN           0   
533           NaN          NaN          NaN      NaN      NaN           0   
534           NaN          NaN          NaN      NaN      NaN           0   
554           NaN          NaN          NaN      NaN      NaN           0   
647           NaN          NaN          NaN      NaN      NaN           0   
706           NaN          NaN          NaN      NaN      NaN           0   
737           NaN          NaN          NaN      NaN      NaN           0   
750           NaN          NaN          NaN      NaN      NaN           0   
779           NaN          NaN          NaN      NaN      NaN           0   
869           NaN          NaN          NaN      NaN      NaN           0   
895           NaN          NaN          NaN      NaN      NaN           0   
898           NaN          NaN          NaN      NaN      NaN           0   
949           NaN          Unf          Unf       TA       Gd           0   
985           NaN          NaN          NaN      NaN      NaN           0   
1001          NaN          NaN          NaN      NaN      NaN           0   
1012          NaN          NaN          NaN      NaN      NaN           0   
1036          NaN          NaN          NaN      NaN      NaN           0   
1046          NaN          NaN          NaN      NaN      NaN           0   
1049          NaN          NaN          NaN      NaN      NaN           0   
1050          NaN          NaN          NaN      NaN      NaN           0   
1091          NaN          NaN          NaN      NaN      NaN           0   
1180          NaN          NaN          NaN      NaN      NaN           0   
1217          NaN          NaN          NaN      NaN      NaN           0   
1219          NaN          NaN          NaN      NaN      NaN           0   
1233          NaN          NaN          NaN      NaN      NaN           0   
1322          NaN          NaN          NaN      NaN      NaN           0   
1413          NaN          NaN          NaN      NaN      NaN           0   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
Id                                        
18             0          0            0  
40             0          0            0  
91             0          0            0  
103            0          0            0  
157            0          0            0  
183            0          0            0  
260            0          0            0  
343            0          0            0  
363            0          0            0  
372            0          0            0  
393            0          0            0  
521            0          0            0  
533            0          0            0  
534            0          0            0  
554            0          0            0  
647            0          0            0  
706            0          0            0  
737            0          0            0  
750            0          0            0  
779            0     

In [41]:
n4 = train.loc[train[['BsmtFinType2']].isnull().all(1)]
n4[bsmt_all_cols]

BsmtExposure BsmtFinType2 BsmtFinType1 BsmtCond BsmtQual  BsmtFinSF1  \
Id                                                                          
18            NaN          NaN          NaN      NaN      NaN           0   
40            NaN          NaN          NaN      NaN      NaN           0   
91            NaN          NaN          NaN      NaN      NaN           0   
103           NaN          NaN          NaN      NaN      NaN           0   
157           NaN          NaN          NaN      NaN      NaN           0   
183           NaN          NaN          NaN      NaN      NaN           0   
260           NaN          NaN          NaN      NaN      NaN           0   
333            No          NaN          GLQ       TA       Gd        1124   
343           NaN          NaN          NaN      NaN      NaN           0   
363           NaN          NaN          NaN      NaN      NaN           0   
372           NaN          NaN          NaN      NaN      NaN           0   
393           NaN          NaN          NaN      NaN      NaN           0   
521           NaN          NaN          NaN      NaN      NaN           0   
533           NaN          NaN          NaN      NaN      NaN           0   
534           NaN          NaN          NaN      NaN      NaN           0   
554           NaN          NaN          NaN      NaN      NaN           0   
647           NaN          NaN          NaN      NaN      NaN           0   
706           NaN          NaN          NaN      NaN      NaN           0   
737           NaN          NaN          NaN      NaN      NaN           0   
750           NaN          NaN          NaN      NaN      NaN           0   
779           NaN          NaN          NaN      NaN      NaN           0   
869           NaN          NaN          NaN      NaN      NaN           0   
895           NaN          NaN          NaN      NaN      NaN           0   
898           NaN          NaN          NaN      NaN      NaN           0   
985           NaN          NaN          NaN      NaN      NaN           0   
1001          NaN          NaN          NaN      NaN      NaN           0   
1012          NaN          NaN          NaN      NaN      NaN           0   
1036          NaN          NaN          NaN      NaN      NaN           0   
1046          NaN          NaN          NaN      NaN      NaN           0   
1049          NaN          NaN          NaN      NaN      NaN           0   
1050          NaN          NaN          NaN      NaN      NaN           0   
1091          NaN          NaN          NaN      NaN      NaN           0   
1180          NaN          NaN          NaN      NaN      NaN           0   
1217          NaN          NaN          NaN      NaN      NaN           0   
1219          NaN          NaN          NaN      NaN      NaN           0   
1233          NaN          NaN          NaN      NaN      NaN           0   
1322          NaN          NaN          NaN      NaN      NaN           0   
1413          NaN          NaN          NaN      NaN      NaN           0   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
Id                                        
18             0          0            0  
40             0          0            0  
91             0          0            0  
103            0          0            0  
157            0          0            0  
183            0          0            0  
260            0          0            0  
333          479       1603         3206  
343            0          0            0  
363            0          0            0  
372            0          0            0  
393            0          0            0  
521            0          0            0  
533            0          0            0  
534            0          0            0  
554            0          0            0  
647            0          0            0  
706            0          0            0  
737            0          0            0  
750            0     

ID 333 has null on only 'BsmtExposure' and ID 949 has null on only 'BsmtFinType2'

In [42]:
joined = n3.merge(n4, left_index=True, right_index=True, how='outer')

In [43]:
joined

MSSubClass_x MSZoning_x  LotArea_x Street_x LotShape_x LandContour_x  \
Id                                                                           
18            90.0         RL    10791.0     Pave        Reg           Lvl   
40            90.0         RL     6040.0     Pave        Reg           Lvl   
91            20.0         RL     7200.0     Pave        Reg           Lvl   
103           90.0         RL     7018.0     Pave        Reg           Bnk   
157           20.0         RL     7200.0     Pave        Reg           Lvl   
183           20.0         RL     9060.0     Pave        Reg           Lvl   
260           20.0         RM    12702.0     Pave        Reg           Lvl   
333            NaN        NaN        NaN      NaN        NaN           NaN   
343           90.0         RL     8544.0     Pave        Reg           Lvl   
363           85.0         RL     7301.0     Pave        Reg           Lvl   
372           50.0         RL    17120.0     Pave        Reg           Lvl   
393           20.0         RL     8339.0     Pave        IR1           Lvl   
521          190.0         RL    10800.0     Pave        Reg           Lvl   
533           20.0         RL     7200.0     Pave        Reg           Lvl   
534           20.0         RL     5000.0     Pave        Reg           Low   
554           20.0         RL     8777.0     Pave        Reg           Lvl   
647           20.0         RL     7200.0     Pave        Reg           Lvl   
706          190.0         RM     5600.0     Pave        Reg           Lvl   
737           90.0         RL     8544.0     Pave        Reg           Lvl   
750           50.0         RL     8405.0     Pave        Reg           Lvl   
779           90.0         RH     8400.0     Pave        Reg           Lvl   
869           60.0         RL    14762.0     Pave        IR2           Lvl   
895           90.0         RL     7018.0     Pave        Reg           Bnk   
898           90.0         RL     7018.0     Pave        Reg           Lvl   
949           60.0         RL    14006.0     Pave        IR1           Lvl   
985           90.0         RL    10125.0     Pave        Reg           Lvl   
1001          20.0         RL    10206.0     Pave        Reg           Lvl   
1012          90.0         RL     9825.0     Pave        Reg           Lvl   
1036          20.0         RL    11500.0     Pave        IR1           Lvl   
1046          20.0         RL    13680.0     Pave        IR1           Lvl   
1049          20.0         RL    21750.0     Pave        Reg           Lvl   
1050          20.0         RL    11100.0     Pave        Reg           Low   
1091          90.0         RL     8544.0     Pave        Reg           Lvl   
1180          20.0         RL     8335.0     Pave        Reg           Lvl   
1217          90.0         RM     8930.0     Pave        Reg           Lvl   
1219          50.0         RM     6240.0     Pave        Reg           Lvl   
1233          90.0         RL     9842.0     Pave        Reg           Lvl   
1322          20.0         RL     6627.0     Pave        IR1           Lvl   
1413          90.0         RL     7200.0     Pave        Reg           Lvl   

     Utilities_x LotConfig_x LandSlope_x Neighborhood_x  ... EnclosedPorch_y  \
Id                                                       ...                   
18        AllPub      Inside         Gtl         Sawyer  ...             0.0   
40        AllPub      Inside         Gtl        Edwards  ...             0.0   
91        AllPub      Inside         Gtl          NAmes  ...             0.0   
103       AllPub      Inside         Gtl        SawyerW  ...             0.0   
157       AllPub      Inside         Gtl          NAmes  ...             0.0   
183       AllPub      Inside         Gtl        Edwards  ...             0.0   
260       AllPub      Inside         Gtl        OldTown  ...             0.0   
333          NaN         NaN         NaN            NaN  ...             0.0   
343    

By outer join, I got 39 rows which have null either 'BsmtExposure' or 'BsmtFinType2', and null on all other 'Bsmt####' rows.

After all, I can replace the null values on 'BsmtExposure','BsmtFinType2', 'BsmtFinType1', 'BsmtCond', and 'BsmtQual' to 'NoBsmt', instead of dropping them.

In [44]:
train = train.fillna({'BsmtExposure':'NoBsmt', 'BsmtFinType2':'NoBsmt', 'BsmtFinType1':'NoBsmt', 'BsmtCond':'NoBsmt', 'BsmtQual':'NoBsmt'})

In [45]:
n = train.loc[train[bsmt_cols].isnull().all(1)]
n[bsmt_all_cols].shape

(0, 9)

Null is all gone in all 5 columns.

### Masonry venner related columns

In [46]:
#Check if there is any abnormal data in 'MasVnrType' column
train['MasVnrType'].value_counts(), train['MasVnrType'].value_counts().sum()

(None       864
 BrkFace    445
 Stone      128
 BrkCmn      15
 Name: MasVnrType, dtype: int64,
 1452)

###### Description
MasVnrType: Masonry veneer type

       BrkCmn	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       None	 None
       Stone	Stone

In [47]:
#Check if there is any abnormal data in 'MasVnrArea' column
train['MasVnrArea'].value_counts(), train['MasVnrArea'].value_counts().sum()

(0.0      861
 72.0       8
 180.0      8
 108.0      8
 120.0      7
         ... 
 651.0      1
 337.0      1
 415.0      1
 293.0      1
 621.0      1
 Name: MasVnrArea, Length: 327, dtype: int64,
 1452)

The null values actually would mean "No masonry veneer area", not really missing data.

In addition, as same as the garage related columns, we can assume that these two columns would have null for the same rows, let's double-check.

In [48]:
mas = ['MasVnrType', 'MasVnrArea']
n = train.loc[train[mas].isnull().all(1)]
n[mas].shape

(8, 2)

In [49]:
mt = train.loc[train['MasVnrType'].isnull()]
mt['MasVnrType'].shape

(8,)

In [50]:
ma = train.loc[train['MasVnrArea'].isnull()]
ma['MasVnrArea'].shape

(8,)

We could see that those columns have the same values for the same records.

I will replace them to "NoMaVe" instead of dropping them.

In [51]:
train = train.fillna({'MasVnrType':'NoMaVe', 'MasVnrArea':0})

In [52]:
n = train.loc[train[mas].isnull().all(1)]
n[mas].shape

(0, 2)

Null is all gone in 2 columns.

### Electrical

In [53]:
#Check if there is any abnormal data in 'Electrical' column
train['Electrical'].value_counts(), train['Electrical'].value_counts().sum()

(SBrkr    1334
 FuseA      94
 FuseF      27
 FuseP       3
 Mix         1
 Name: Electrical, dtype: int64,
 1459)

###### Description

Electrical: Electrical system

       SBrkr	Standard Circuit Breakers & Romex
       FuseA	Fuse Box over 60 AMP and all Romex wiring (Average)	
       FuseF	60 AMP Fuse Box and mostly Romex wiring (Fair)
       FuseP	60 AMP Fuse Box and mostly knob & tube wiring (poor)
       Mix	Mixed

In [54]:
#Find out which row has the null
n = train.loc[train['Electrical'].isnull()]
n['Electrical']

Id
1380    NaN
Name: Electrical, dtype: object

In [55]:
train.iloc[1377:1382, 37:42]

HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF
Id                                                      
1378        Ex          Y      SBrkr       984       620
1379        TA          Y      SBrkr       483       504
1380        Gd          Y        NaN       754       640
1381        TA          N      FuseF       864         0
1382        Ex          Y      SBrkr      2117         0

I will replace the null with 'SBrkr' as the majority of data have it.

In [56]:
train = train.fillna({'Electrical':'SBrkr'})
train.iloc[1377:1382, 37:42]

HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF
Id                                                      
1378        Ex          Y      SBrkr       984       620
1379        TA          Y      SBrkr       483       504
1380        Gd          Y      SBrkr       754       640
1381        TA          N      FuseF       864         0
1382        Ex          Y      SBrkr      2117         0

The null value in 'Electrical' for ID 1380 has been replaced.

## Other columns having no missing value

Check if each column has abnormal value or outlier refering to 'data_description.txt' file.

In [57]:
train['MSSubClass'].value_counts().sort_values(ascending=False), train['MSSubClass'].value_counts().sum()

(20     536
 60     299
 50     144
 120     87
 30      69
 160     63
 70      60
 80      58
 90      52
 190     30
 85      20
 75      16
 45      12
 180     10
 40       4
 Name: MSSubClass, dtype: int64,
 1460)

In [58]:
train['MSZoning'].value_counts().sort_values(ascending=False), train['MSZoning'].value_counts().sum()

(RL         1151
 RM          218
 FV           65
 RH           16
 C (all)      10
 Name: MSZoning, dtype: int64,
 1460)

In [59]:
train['LotArea'].value_counts().sort_values(ascending=False), train['LotArea'].value_counts().sum()

(7200     25
 9600     24
 6000     17
 10800    14
 9000     14
          ..
 13682     1
 9587      1
 9588      1
 9590      1
 8123      1
 Name: LotArea, Length: 1073, dtype: int64,
 1460)

In [60]:
train['Street'].value_counts().sort_values(ascending=False), train['Street'].value_counts().sum()

(Pave    1454
 Grvl       6
 Name: Street, dtype: int64,
 1460)

'Street' column can be dropped as the most values are 'Pave'.

In [61]:
train = train.drop(['Street'], axis=1)

In [62]:
print(train.shape)

(1460, 73)


In [63]:
train['LotShape'].value_counts().sort_values(ascending=False), train['LotShape'].value_counts().sum()

(Reg    925
 IR1    484
 IR2     41
 IR3     10
 Name: LotShape, dtype: int64,
 1460)

In [64]:
train['LandContour'].value_counts().sort_values(ascending=False), train['LandContour'].value_counts().sum()

(Lvl    1311
 Bnk      63
 HLS      50
 Low      36
 Name: LandContour, dtype: int64,
 1460)

In [65]:
train['Utilities'].value_counts().sort_values(ascending=False), train['Utilities'].value_counts().sum()

(AllPub    1459
 NoSeWa       1
 Name: Utilities, dtype: int64,
 1460)

'Utilities' column can be dropped as the most values are 'AllPub'.

In [66]:
train = train.drop(['Utilities'], axis=1)

In [67]:
print(train.shape)

(1460, 72)


In [68]:
train['LotConfig'].value_counts().sort_values(ascending=False), train['LotConfig'].value_counts().sum()

(Inside     1052
 Corner      263
 CulDSac      94
 FR2          47
 FR3           4
 Name: LotConfig, dtype: int64,
 1460)

In [69]:
train['LandSlope'].value_counts().sort_values(ascending=False), train['LandSlope'].value_counts().sum()

(Gtl    1382
 Mod      65
 Sev      13
 Name: LandSlope, dtype: int64,
 1460)

In [70]:
train['Neighborhood'].value_counts().sort_values(ascending=False), train['Neighborhood'].value_counts().sum()

(NAmes      225
 CollgCr    150
 OldTown    113
 Edwards    100
 Somerst     86
 Gilbert     79
 NridgHt     77
 Sawyer      74
 NWAmes      73
 SawyerW     59
 BrkSide     58
 Crawfor     51
 Mitchel     49
 NoRidge     41
 Timber      38
 IDOTRR      37
 ClearCr     28
 SWISU       25
 StoneBr     25
 Blmngtn     17
 MeadowV     17
 BrDale      16
 Veenker     11
 NPkVill      9
 Blueste      2
 Name: Neighborhood, dtype: int64,
 1460)

In [71]:
train['Condition1'].value_counts().sort_values(ascending=False), train['Condition1'].value_counts().sum()

(Norm      1260
 Feedr       81
 Artery      48
 RRAn        26
 PosN        19
 RRAe        11
 PosA         8
 RRNn         5
 RRNe         2
 Name: Condition1, dtype: int64,
 1460)

In [72]:
train['Condition2'].value_counts().sort_values(ascending=False), train['Condition2'].value_counts().sum()

(Norm      1445
 Feedr        6
 Artery       2
 RRNn         2
 PosN         2
 PosA         1
 RRAn         1
 RRAe         1
 Name: Condition2, dtype: int64,
 1460)

'Condition2' column can be dropped as the most values are 'Norm'.

In [73]:
train = train.drop(['Condition2'], axis=1)

In [74]:
print(train.shape)

(1460, 71)


In [75]:
train['BldgType'].value_counts().sort_values(ascending=False), train['BldgType'].value_counts().sum()

(1Fam      1220
 TwnhsE     114
 Duplex      52
 Twnhs       43
 2fmCon      31
 Name: BldgType, dtype: int64,
 1460)

In [76]:
train['HouseStyle'].value_counts().sort_values(ascending=False), train['HouseStyle'].value_counts().sum()

(1Story    726
 2Story    445
 1.5Fin    154
 SLvl       65
 SFoyer     37
 1.5Unf     14
 2.5Unf     11
 2.5Fin      8
 Name: HouseStyle, dtype: int64,
 1460)

In [77]:
train['OverallQual'].value_counts().sort_values(ascending=False), train['OverallQual'].value_counts().sum()

(5     397
 6     374
 7     319
 8     168
 4     116
 9      43
 3      20
 10     18
 2       3
 1       2
 Name: OverallQual, dtype: int64,
 1460)

In [78]:
train['OverallCond'].value_counts().sort_values(ascending=False), train['OverallCond'].value_counts().sum()

(5    821
 6    252
 7    205
 8     72
 4     57
 3     25
 9     22
 2      5
 1      1
 Name: OverallCond, dtype: int64,
 1460)

In [79]:
train['YearBuilt'].value_counts().sort_values(ascending=False), train['YearBuilt'].value_counts().sum()

(2006    67
 2005    64
 2004    54
 2007    49
 2003    45
         ..
 2010     1
 1893     1
 1875     1
 1882     1
 1872     1
 Name: YearBuilt, Length: 112, dtype: int64,
 1460)

In [80]:
train['YearRemodAdd'].value_counts().sort_values(ascending=False), train['YearRemodAdd'].value_counts().sum()

(1950    178
 2006     97
 2007     76
 2005     73
 2004     62
        ... 
 2010      6
 1952      5
 1986      5
 1983      5
 1951      4
 Name: YearRemodAdd, Length: 61, dtype: int64,
 1460)

In [81]:
train['RoofStyle'].value_counts().sort_values(ascending=False), train['RoofStyle'].value_counts().sum()

(Gable      1141
 Hip         286
 Flat         13
 Gambrel      11
 Mansard       7
 Shed          2
 Name: RoofStyle, dtype: int64,
 1460)

In [82]:
train['RoofMatl'].value_counts().sort_values(ascending=False), train['RoofMatl'].value_counts().sum()

(CompShg    1434
 Tar&Grv      11
 WdShngl       6
 WdShake       5
 ClyTile       1
 Membran       1
 Roll          1
 Metal         1
 Name: RoofMatl, dtype: int64,
 1460)

'RoofMatl' column can be dropped as the most values are 'CompShg'.

In [83]:
train = train.drop(['RoofMatl'], axis=1)

In [84]:
print(train.shape)

(1460, 70)


In [85]:
train['Exterior1st'].value_counts().sort_values(ascending=False), train['Exterior1st'].value_counts().sum()

(VinylSd    515
 HdBoard    222
 MetalSd    220
 Wd Sdng    206
 Plywood    108
 CemntBd     61
 BrkFace     50
 WdShing     26
 Stucco      25
 AsbShng     20
 Stone        2
 BrkComm      2
 AsphShn      1
 ImStucc      1
 CBlock       1
 Name: Exterior1st, dtype: int64,
 1460)

In [86]:
train['Exterior2nd'].value_counts().sort_values(ascending=False), train['Exterior2nd'].value_counts().sum()

(VinylSd    504
 MetalSd    214
 HdBoard    207
 Wd Sdng    197
 Plywood    142
 CmentBd     60
 Wd Shng     38
 Stucco      26
 BrkFace     25
 AsbShng     20
 ImStucc     10
 Brk Cmn      7
 Stone        5
 AsphShn      3
 CBlock       1
 Other        1
 Name: Exterior2nd, dtype: int64,
 1460)

In [87]:
train['ExterQual'].value_counts().sort_values(ascending=False), train['ExterQual'].value_counts().sum()

(TA    906
 Gd    488
 Ex     52
 Fa     14
 Name: ExterQual, dtype: int64,
 1460)

In [88]:
train['ExterCond'].value_counts().sort_values(ascending=False), train['ExterCond'].value_counts().sum()

(TA    1282
 Gd     146
 Fa      28
 Ex       3
 Po       1
 Name: ExterCond, dtype: int64,
 1460)

In [89]:
train['Foundation'].value_counts().sort_values(ascending=False), train['Foundation'].value_counts().sum()

(PConc     647
 CBlock    634
 BrkTil    146
 Slab       24
 Stone       6
 Wood        3
 Name: Foundation, dtype: int64,
 1460)

In [90]:
train['BsmtFinSF1'].value_counts().sort_values(ascending=False), train['BsmtFinSF1'].value_counts().sum()

(0       467
 24       12
 16        9
 616       5
 936       5
        ... 
 1441      1
 224       1
 223       1
 222       1
 404       1
 Name: BsmtFinSF1, Length: 637, dtype: int64,
 1460)

In [91]:
train['BsmtFinSF2'].value_counts().sort_values(ascending=False), train['BsmtFinSF2'].value_counts().sum()

(0      1293
 180       5
 374       3
 279       2
 469       2
        ... 
 352       1
 764       1
 713       1
 692       1
 169       1
 Name: BsmtFinSF2, Length: 144, dtype: int64,
 1460)

In [92]:
train['BsmtUnfSF'].value_counts().sort_values(ascending=False), train['BsmtUnfSF'].value_counts().sum()

(0       118
 728       9
 384       8
 572       7
 600       7
        ... 
 1078      1
 1079      1
 1082      1
 1085      1
 1777      1
 Name: BsmtUnfSF, Length: 780, dtype: int64,
 1460)

In [93]:
train['TotalBsmtSF'].value_counts().sort_values(ascending=False), train['TotalBsmtSF'].value_counts().sum()

(0       37
 864     35
 672     17
 912     15
 1040    14
         ..
 1926     1
 662      1
 1842     1
 1766     1
 1614     1
 Name: TotalBsmtSF, Length: 721, dtype: int64,
 1460)

In [94]:
train['Heating'].value_counts().sort_values(ascending=False), train['Heating'].value_counts().sum()

(GasA     1428
 GasW       18
 Grav        7
 Wall        4
 OthW        2
 Floor       1
 Name: Heating, dtype: int64,
 1460)

'Heating' column can be dropped as the most values are 'GasA'.

In [95]:
train = train.drop(['Heating'], axis=1)

In [96]:
print(train.shape)

(1460, 69)


In [97]:
train['HeatingQC'].value_counts().sort_values(ascending=False), train['HeatingQC'].value_counts().sum()

(Ex    741
 TA    428
 Gd    241
 Fa     49
 Po      1
 Name: HeatingQC, dtype: int64,
 1460)

In [98]:
train['CentralAir'].value_counts().sort_values(ascending=False), train['CentralAir'].value_counts().sum()

(Y    1365
 N      95
 Name: CentralAir, dtype: int64,
 1460)

In [99]:
train['1stFlrSF'].value_counts().sort_values(ascending=False), train['1stFlrSF'].value_counts().sum()

(864     25
 1040    16
 912     14
 848     12
 894     12
         ..
 1586     1
 1582     1
 1578     1
 1576     1
 969      1
 Name: 1stFlrSF, Length: 753, dtype: int64,
 1460)

In [100]:
train['2ndFlrSF'].value_counts().sort_values(ascending=False), train['2ndFlrSF'].value_counts().sum()

(0       829
 728      10
 504       9
 672       8
 546       8
        ... 
 1037      1
 1031      1
 1028      1
 892       1
 430       1
 Name: 2ndFlrSF, Length: 417, dtype: int64,
 1460)

In [101]:
train['LowQualFinSF'].value_counts().sort_values(ascending=False), train['LowQualFinSF'].value_counts().sum()

(0      1434
 80        3
 360       2
 515       1
 528       1
 53        1
 120       1
 144       1
 156       1
 205       1
 232       1
 234       1
 384       1
 572       1
 390       1
 392       1
 397       1
 420       1
 473       1
 479       1
 481       1
 513       1
 514       1
 371       1
 Name: LowQualFinSF, dtype: int64,
 1460)

'LowQualFinSF' column can be dropped as the most values are 0.

In [102]:
train = train.drop(['LowQualFinSF'], axis=1)

In [103]:
print(train.shape)

(1460, 68)


In [104]:
train['GrLivArea'].value_counts().sort_values(ascending=False), train['GrLivArea'].value_counts().sum()

(864     22
 1040    14
 894     11
 848     10
 1456    10
         ..
 1787     1
 1797     1
 1764     1
 1803     1
 2230     1
 Name: GrLivArea, Length: 861, dtype: int64,
 1460)

In [105]:
train['BsmtFullBath'].value_counts().sort_values(ascending=False), train['BsmtFullBath'].value_counts().sum()

(0    856
 1    588
 2     15
 3      1
 Name: BsmtFullBath, dtype: int64,
 1460)

In [106]:
train['BsmtHalfBath'].value_counts().sort_values(ascending=False), train['BsmtHalfBath'].value_counts().sum()

(0    1378
 1      80
 2       2
 Name: BsmtHalfBath, dtype: int64,
 1460)

In [107]:
train['FullBath'].value_counts().sort_values(ascending=False), train['FullBath'].value_counts().sum()

(2    768
 1    650
 3     33
 0      9
 Name: FullBath, dtype: int64,
 1460)

In [108]:
train['HalfBath'].value_counts().sort_values(ascending=False), train['HalfBath'].value_counts().sum()

(0    913
 1    535
 2     12
 Name: HalfBath, dtype: int64,
 1460)

In [109]:
train['BedroomAbvGr'].value_counts().sort_values(ascending=False), train['BedroomAbvGr'].value_counts().sum()

(3    804
 2    358
 4    213
 1     50
 5     21
 6      7
 0      6
 8      1
 Name: BedroomAbvGr, dtype: int64,
 1460)

In [110]:
train['KitchenAbvGr'].value_counts().sort_values(ascending=False), train['KitchenAbvGr'].value_counts().sum()

(1    1392
 2      65
 3       2
 0       1
 Name: KitchenAbvGr, dtype: int64,
 1460)

In [111]:
train['KitchenQual'].value_counts().sort_values(ascending=False), train['KitchenQual'].value_counts().sum()

(TA    735
 Gd    586
 Ex    100
 Fa     39
 Name: KitchenQual, dtype: int64,
 1460)

In [112]:
train['TotRmsAbvGrd'].value_counts().sort_values(ascending=False), train['TotRmsAbvGrd'].value_counts().sum()

(6     402
 7     329
 5     275
 8     187
 4      97
 9      75
 10     47
 11     18
 3      17
 12     11
 2       1
 14      1
 Name: TotRmsAbvGrd, dtype: int64,
 1460)

In [113]:
train['Functional'].value_counts().sort_values(ascending=False), train['Functional'].value_counts().sum()

(Typ     1360
 Min2      34
 Min1      31
 Mod       15
 Maj1      14
 Maj2       5
 Sev        1
 Name: Functional, dtype: int64,
 1460)

In [114]:
train['Fireplaces'].value_counts().sort_values(ascending=False), train['Fireplaces'].value_counts().sum()

(0    690
 1    650
 2    115
 3      5
 Name: Fireplaces, dtype: int64,
 1460)

In [115]:
train['GarageCars'].value_counts().sort_values(ascending=False), train['GarageCars'].value_counts().sum()

(2    824
 1    369
 3    181
 0     81
 4      5
 Name: GarageCars, dtype: int64,
 1460)

In [116]:
train['GarageArea'].value_counts().sort_values(ascending=False), train['GarageArea'].value_counts().sum()

(0       81
 440     49
 576     47
 240     38
 484     34
         ..
 319      1
 261      1
 318      1
 309      1
 1418     1
 Name: GarageArea, Length: 441, dtype: int64,
 1460)

In [117]:
train['PavedDrive'].value_counts().sort_values(ascending=False), train['PavedDrive'].value_counts().sum()

(Y    1340
 N      90
 P      30
 Name: PavedDrive, dtype: int64,
 1460)

In [118]:
train['WoodDeckSF'].value_counts().sort_values(ascending=False), train['WoodDeckSF'].value_counts().sum()

(0      761
 192     38
 100     36
 144     33
 120     31
       ... 
 406      1
 409      1
 414      1
 416      1
 35       1
 Name: WoodDeckSF, Length: 274, dtype: int64,
 1460)

In [119]:
train['OpenPorchSF'].value_counts().sort_values(ascending=False), train['OpenPorchSF'].value_counts().sum()

(0      656
 36      29
 48      22
 20      21
 40      19
       ... 
 224      1
 229      1
 231      1
 235      1
 119      1
 Name: OpenPorchSF, Length: 202, dtype: int64,
 1460)

In [120]:
train['EnclosedPorch'].value_counts().sort_values(ascending=False), train['EnclosedPorch'].value_counts().sum()

(0      1252
 112      15
 96        6
 120       5
 144       5
        ... 
 286       1
 218       1
 220       1
 226       1
 37        1
 Name: EnclosedPorch, Length: 120, dtype: int64,
 1460)

In [121]:
train['3SsnPorch'].value_counts().sort_values(ascending=False), train['3SsnPorch'].value_counts().sum()

(0      1436
 168       3
 216       2
 144       2
 180       2
 23        1
 245       1
 238       1
 290       1
 196       1
 508       1
 407       1
 304       1
 162       1
 153       1
 320       1
 140       1
 130       1
 96        1
 182       1
 Name: 3SsnPorch, dtype: int64,
 1460)

'3SsnPorch' column can be dropped as the most values are 0.

In [122]:
train = train.drop(['3SsnPorch'], axis=1)

In [123]:
print(train.shape)

(1460, 67)


In [124]:
train['ScreenPorch'].value_counts().sort_values(ascending=False), train['ScreenPorch'].value_counts().sum()

(0      1344
 192       6
 224       5
 120       5
 189       4
        ... 
 273       1
 271       1
 266       1
 265       1
 99        1
 Name: ScreenPorch, Length: 76, dtype: int64,
 1460)

In [125]:
train['PoolArea'].value_counts().sort_values(ascending=False), train['PoolArea'].value_counts().sum()

(0      1453
 480       1
 512       1
 519       1
 555       1
 576       1
 648       1
 738       1
 Name: PoolArea, dtype: int64,
 1460)

'PoolArea' column can be dropped as the most values are 0.

In [126]:
train = train.drop(['PoolArea'], axis=1)

In [127]:
print(train.shape)

(1460, 66)


In [128]:
train['MiscVal'].value_counts().sort_values(ascending=False), train['MiscVal'].value_counts().sum()

(0        1408
 400        11
 500         8
 700         5
 450         4
 2000        4
 600         4
 1200        2
 480         2
 1150        1
 54          1
 8300        1
 15500       1
 620         1
 3500        1
 560         1
 2500        1
 1300        1
 1400        1
 350         1
 800         1
 Name: MiscVal, dtype: int64,
 1460)

As 'MiscFeature' has been dropped at the beginning, 'MiscVal' is useless and dropping as well.

In [129]:
train = train.drop(['MiscVal'], axis=1)

In [130]:
print(train.shape)

(1460, 65)


In [131]:
train['MoSold'].value_counts().sort_values(ascending=False), train['MoSold'].value_counts().sum()

(6     253
 7     234
 5     204
 4     141
 8     122
 3     106
 10     89
 11     79
 9      63
 12     59
 1      58
 2      52
 Name: MoSold, dtype: int64,
 1460)

In [132]:
train['YrSold'].value_counts().sort_values(ascending=False), train['YrSold'].value_counts().sum()

(2009    338
 2007    329
 2006    314
 2008    304
 2010    175
 Name: YrSold, dtype: int64,
 1460)

In [133]:
train['SaleType'].value_counts().sort_values(ascending=False), train['SaleType'].value_counts().sum()

(WD       1267
 New       122
 COD        43
 ConLD       9
 ConLI       5
 ConLw       5
 CWD         4
 Oth         3
 Con         2
 Name: SaleType, dtype: int64,
 1460)

In [134]:
train['SaleCondition'].value_counts().sort_values(ascending=False), train['SaleCondition'].value_counts().sum()

(Normal     1198
 Partial     125
 Abnorml     101
 Family       20
 Alloca       12
 AdjLand       4
 Name: SaleCondition, dtype: int64,
 1460)

In [135]:
#Check a summary of the data again after dropped some columns
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 65 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1460 non-null   int64  
 1   MSZoning       1460 non-null   object 
 2   LotArea        1460 non-null   int64  
 3   LotShape       1460 non-null   object 
 4   LandContour    1460 non-null   object 
 5   LotConfig      1460 non-null   object 
 6   LandSlope      1460 non-null   object 
 7   Neighborhood   1460 non-null   object 
 8   Condition1     1460 non-null   object 
 9   BldgType       1460 non-null   object 
 10  HouseStyle     1460 non-null   object 
 11  OverallQual    1460 non-null   int64  
 12  OverallCond    1460 non-null   int64  
 13  YearBuilt      1460 non-null   int64  
 14  YearRemodAdd   1460 non-null   int64  
 15  RoofStyle      1460 non-null   object 
 16  Exterior1st    1460 non-null   object 
 17  Exterior2nd    1460 non-null   object 
 18  MasVnrTy

# Things to do

- 'GarageYrBlt' needs to be transited to date format

- All square feet data columns has too many unique rows, how to deal with them?

# Idea for the next

- Drop 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2' and 'BsmtUnfSF' as 'TotalBsmtSF' would be enough for buyser in general.

- Sum 'BsmtFullBath' & 'FullBath' / 'BsmtHalfBath' & 'HalfBath' as people usually consider the total bathrooms / half bathroom in the property, not by level.

- Drop 'BedroomAbvGr' as 'TotRmsAbvGrd' would be more useful in general


